pina/geometry/difference_domain.py

```diff
@@ -1,28 +1,142 @@
+        """Check if a point is inside the difference domain.
+
+        :param point: Point to be checked.
+        :type point: torch.Tensor   
+        :param bool check_border: If True, the border is considered inside.
+        :return: True if the point is inside the difference domain, False otherwise.
+        :rtype: bool
+        """
+        flag = 0
+        for geometry in self.geometries:
+            if geometry.is_inside(point, check_border):
+                flag += 1
+        return flag == 1
+
+    def sample(self, n, mode='random', variables='all'):
+        """Sample routine.
 
-        import torch
-        return LabelTensor(torch.cat(samples), labels=['x', 'y'])
+        :param n: Number of points to sample in the shape.
+        :type n: int
+        :param mode: Mode for sampling, defaults to 'random'.
+            Available modes include: random sampling, 'random'.
+        :type mode: str, optional
+        :param variables: pinn variable to be sampled, defaults to 'all'.
+        :type variables: str or list[str], optional
+
+        :Example:
+            # Create two Cartesian domains
+            >>> cartesian1 = CartesianDomain({'x': [0, 2], 'y': [0, 2]})
+            >>> cartesian2 = CartesianDomain({'x': [1, 3], 'y': [1, 3]})
+
+            # Create a difference of the ellipsoid domains
+            >>> difference = Difference([cartesian1, cartesian2])
+
+            >>> difference.sample(n=1000)
+                LabelTensor([[1.6802, 0.7723],
+                    [1.8872, 0.9972],
+                    [0.3166, 0.1213],
+                    ...,
+                    [2.1999, 2.0767],
+                    [1.0655, 2.7711],
+                    [1.6751, 2.3695]])
+
+            >>> len(difference.sample(n=1000)
+                1000
+
+        """
+        sampled = []
+
+        # calculate the number of points to sample for each geometry and the remainder.
+        remainder = n % len(self.geometries)
+        num_points = n // len(self.geometries)
+
+        # sample the points
+        # NB. geometries as shuffled since if we sample
+        # multiple times just one point, we would end
+        # up sampling only from the first geometry.
+        iter_ = random.sample(self.geometries, len(self.geometries))
+        for i, geometry in enumerate(iter_):
+            sampled_points = []
+            # int(i < remainder) is one only if we have a remainder
+            # different than zero. Notice that len(geometries) is
+            # always smaller than remaider.
+            # makes sure point is uniquely inside 1 shape.
+            while len(sampled_points) < (num_points + int(i < remainder)):
+                sample = geometry.sample(1, 'random')
+                # if not self.is_inside(sample) --> will be the intersection
+                if self.is_inside(sample):
+                    sampled_points.append(sample)
+            sampled.extend(sampled_points)
```

Comment:
Can you explain me by answering this comment what is your strategy for sampling?

Reply:
`sample = geometry.sampled(1, 'random')` samples 1 point from the current geometry using that geometry's own `.sample` method in the 'random' mode. I could change this such that it changes depending on the mode the user wants to use but at the moment it's hard-coded to be "random". 

Then, `if self.is_inside(sample):` checks to see if the sampled point, `sample`, is inside this difference geometry. It does this by using the Difference's `is_inside` method. It checks to see if this sampled point is inside only 1 of the geometries that make up the difference. Effectively, this sampled point should only be inside the geometry of the point that is being sampled (I haven't done a deep dive into how sampling works yet but it's my understanding that for example `CartesianDomain`'s `.sample()` method will return a point that's inside the defined `CartesianDomain`). So, if this is the case, then `if self.is_inside(sample)` is `True`. Then I add the point to `sampled_points`, a list created to add points to keep track of all the points sampled for a certain geometry.* I keep adding until I have `num_points + int(i < remainder` points that have been sampled. I then add this final list of points to `sample`, the final list of points.

Then I do to the next geometry and repeat the process.

*As I write up this explanation, I realize now that I might want to make `sampled_points` a set to ensure that I have unique points because I'm not actually checking to ensure that.

Reply:
Ok so everything should work properly. You don't need a set, it's very unlikely that in random mode you are going to get two points the same